### PR TITLE
Hide Content Type filter when using Admin Menus Content Type feature

### LINKED ISSUE #7094: Hide contenttype filter when a user list elements of a specific content type
In according on what write in #7017, I open this new issue about the need to hide contenttype filter when a user click on an AdminMenu link that send to an interface where are listed the items related to a specific contenttype. 
pr_contenttype_filter_gone

But when a user click on "Content Items" link, the contenttype filter return visible.
pr_contenttype_filter_visible

### PR DESCRIPTION
**Pull request by PiemP**
Fixes #7094 

When contentTypeId have a value means the user want to list the items of a specific content type and the ContentType filter is not necessary. I have left the possiblity to set up ContentTypeOptions from the query string.

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs [hunk 1]
```diff
@@ -156,32 +156,40 @@ public async Task<IActionResult> List(ListContentsViewModel model, PagerParamete
...
+                    new SelectListItem() { Text = S["All content types"], Value = "" }
+                };
+                foreach (var option in contentTypeOptions)
+                {
+                    model.Options.ContentTypeOptions.Add(new SelectListItem() { Text = option.Value, Value = option.Key });
+                }
             }
 
+            //if ContentTypeOptions is not initialized by query string or by the code above, initialize it
+            if (model.Options.ContentTypeOptions == null)
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs:190]: Formatting (braces)

File: src/OrchardCore.Modules/OrchardCore.Contents/Views/ContentsAdminList-Filters.cshtml [hunk 2]
```diff
@@ -4,8 +4,10 @@
 }
 <div class="form-group mb-n1 filter">
     <div class="btn-group mt-1">
-        <select asp-for="SelectedContentType" asp-items="Model.ContentTypeOptions" class="selectpicker contenttypes-selectpicker show-tick mr-2" data-header="@T["Filter by content type"]" data-live-search="true" data-selected-text-format="static" data-width="fit" title="@T["Content Type"]" data-style="btn-sm">
-        </select>
+        @if(Model.ContentTypeOptions.Any()){
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Contents/Views/ContentsAdminList-Filters.cshtml:7]: formatting new line braces

File: src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs [hunk 3]
```diff
@@ -156,32 +156,40 @@ public async Task<IActionResult> List(ListContentsViewModel model, PagerParamete
                 new SelectListItem() { Text = S["Delete"], Value = nameof(ContentsBulkAction.Remove) }
             };
 
-            var listableTypes = new List<ContentTypeDefinition>();
-            foreach (var ctd in _contentDefinitionManager.ListTypeDefinitions())
+            //doesn't fill model.Options.ContentTypeOptions to hide "Content Type" filter on the right side of the interface, because user viewing a list of a specific contenttype items
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs:159]:
```suggestion
            // When using the AdminMenus ContentTypes Feature and specifying a Content Type hide the 'Content Type' filter.
```

File: src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs [hunk 4]
```diff
@@ -156,32 +156,40 @@ public async Task<IActionResult> List(ListContentsViewModel model, PagerParamete
...
+            if (string.IsNullOrEmpty(contentTypeId) && model.Options.ContentTypeOptions == null)
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs:160]:
```suggestion
            if (String.IsNullOrEmpty(contentTypeId) && model.Options.ContentTypeOptions == null)
```

File: src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs (hunk 1 above)
Comment (PiemP) [reply src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs:190]: added some braces.

File: src/OrchardCore.Modules/OrchardCore.Contents/Views/ContentsAdminList-Filters.cshtml (hunk 2 above)
Comment (PiemP) [reply src/OrchardCore.Modules/OrchardCore.Contents/Views/ContentsAdminList-Filters.cshtml]: moved brace to new line

**Review by deanmarcussen — APPROVED** [review]

**Comment by Skrypt** [comment]
I have an issue with this. It doesn't remove the filter only when we are using an admin menu that points to a specific Content Type. It also removes the filter when selecting a Content Type option `Content Items` page.

**Comment by PiemP** [comment]
@Skrypt sorry (really really sorry) but I can't repro the problem. Please, give me some step to reprocude it.
Sorry (I'm felling dumb).

check_content_type_filter

I have tested it with blog reciper and then I have manually create a link but I can't reproduce it:
check_content_type_filter_2

**Comment by deanmarcussen** [comment]
@PiemP I can repro the same issue as @Skrypt 

Latest dev branch.

Is it possible you have a bad merge on your site (because you build from source code I believe?), or the code that went into OC dev is different to what you have?

**Comment by PiemP** [comment]
@deanmarcussen it's a fork of OC, dev branch with no modification, updated 11 hour ago (last PR merged 7311), executed by Visual studio code in debug mode and site initilialized from 0 with Blog recipe.

I have done a `dotnet clean` and test it another time but I can't repro it.

The first gif uploaded show the functionality with the "standard" content items link, the second gif show a custom link created with the Admin Menu interface that allow to list all the content items.

I would try to understand where is the problem, for now I suppose the change in the permission evaluation could break the original behavior.

Sorry if I make you loosing you time.

**Comment by Skrypt** [comment]
I reverted those changes in https://github.com/OrchardCMS/OrchardCore/pull/7299/files
I don't think permission will make any difference on this. The condition is on Model.SelectedContentType which is passed from the Request.Querystring. We can't make a distinction between actually filtering from the UI because it redirects by doing a full GET request to the same page. So we redirect to admin/contentitem/{contenttypename} on both cases. The proper fix would be to use a form post on the filtering component in the admin UI so that we can make a different action when we POST or GET.

**Comment by PiemP** [comment]
Honestly I don't understand how it happens and it's difficult for me explain in English.

To hide the filter `model.Options.ContentTypeOptions` should be null until line 213 of the `AdminController.cs` file. To obtain this, the `contentTypeId` have a value (and this should be happen when a user try to list elements of a specific content type and, because is try to list items of a specific content type, he doesn't need the filter). A user should can't set `contentTypeId` by modify the filter in the **Content Items** interface: this means the value selected in the filter goes in some way in the `contentTypeId` parameter. But this should be prevented by this code:

https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Contents/AdminMenu.cs#L44

https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Contents/Views/ContentsAdminList-Filters.cshtml#L3

Unless someone put the value in the URL direcly like:
##Link Content Items normal url
https://localhost:5001/Admin/Contents/ContentItems?Options.OrderBy=Modified&Options.ContentsStatus=Latest
##Link to Content Item url with `contentTypeId`      
https://localhost:5001/Admin/Contents/ContentItems/Page?Options.OrderBy=Modified&Options.ContentsStatus=Latest

The filter shouldn't be disappear. Sure, passing parameters by a POST request avoid all this problems.

But maybe I have find something that could be related to the @Skrypt issue. This codes seems change the form action and set the `contentTypeId` parameter in URL: 

https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Contents/Views/ContentsAdminList-Filters.cshtml#L17-L31

**Comment by Skrypt** [comment]
I think the issue is with the web component doing a GET request instead of a POST. This is a form and it should POST to the same controller with a different handling of the model.Options.

**Pull request merged**